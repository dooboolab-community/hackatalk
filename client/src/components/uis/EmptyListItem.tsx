import {StyleSheet, Text, View, ViewStyle} from 'react-native';

import React from 'react';

interface Styles {
  wrapper: ViewStyle;
}

const styles: Styles = StyleSheet.create({
  wrapper: {
    flex: 1,
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'stretch',
  },
});

interface Props {
  style: ViewStyle;
  children?: string;
}

function Shared(props: Props): React.ReactElement {
  return (
    <View style={props.style}>
      <Text
        style={{
          fontSize: 14,
          color: 'rgb(155,155,155)',
          alignSelf: 'center',
        }}>
        {props.children}
      </Text>
    </View>
  );
}

Shared.defaultProps = {
  style: styles.wrapper,
};

export default Shared;
