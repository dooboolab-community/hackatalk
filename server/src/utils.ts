import { Context } from './context';
import path from 'path';
import { verify } from 'jsonwebtoken';

const env = process.env.NODE_ENV;
const envPath = env === 'development'
  ? path.resolve(__dirname, '../dotenv/dev.env')
  : env === 'test'
    ? path.resolve(__dirname, '../dotenv/test.env')
    : path.resolve(__dirname, '../dotenv/.env');

require('dotenv').config({ path: envPath });

const { JWT_SECRET } = process.env;
export const APP_SECRET = JWT_SECRET;

interface Token {
  userId: string;
}

export function getUserId(context: Context): string {
  const Authorization = context.request.req.get('Authorization');
  if (Authorization) {
    const token = Authorization.replace('Bearer ', '');
    const verifiedToken = verify(token, APP_SECRET) as Token;
    return verifiedToken && verifiedToken.userId;
  }
}
