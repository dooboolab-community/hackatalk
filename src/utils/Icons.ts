import icAddW from '../../assets/icons/btnAdd.png';
import icBack from '../../assets/icons/btnBack.png';
import icCamera from '../../assets/icons/camera.png';
import icCircleX from '../../assets/icons/x_circle.png';
import icError from '../../assets/icons/error.png';
import icIcon from '../../assets/icons/icon.png';
import icLogoD from '../../assets/icons/logo_d.png';
import icLogoW from '../../assets/icons/logo_w.png';
import icPic from '../../assets/icons/pictoGallery.png';
import icProfile from '../../assets/icons/profile.png';
import icProfileW from '../../assets/icons/profile_w.png';
import icSearch from '../../assets/icons/search.png';
import icSearchW from '../../assets/icons/search_w.png';
import icSettingW from '../../assets/icons/setting_w.png';
import icSmile from '../../assets/icons/icSmile.png';
import svgApple from '../../assets/icons/apple.svg';
import svgFacebook from '../../assets/icons/facebook.svg';
import svgGoogle from '../../assets/icons/google.svg';
import svgPlus from '../../assets/icons/plus.svg';

export const IC_ADD_W = icAddW;
export const IC_ICON = icIcon;
export const IC_CAMERA = icCamera;
export const IC_CIRCLE_X = icCircleX;
export const IC_ERROR = icError;
export const IC_PROFILE = icProfile;
export const IC_BACK = icBack;
export const IC_PIC = icPic;
export const IC_SEARCH = icSearch;
export const IC_SMILE = icSmile;
export const IC_LOGO_D = icLogoD;
export const IC_LOGO_W = icLogoW;
export const IC_SETTING_W = icSettingW;
export const IC_SEARCH_W = icSearchW;
export const IC_PROFILE_W = icProfileW;
export const SvgApple = svgApple;
export const SvgGoogle = svgGoogle;
export const SvgFacebook = svgFacebook;
export const SvgPlus = svgPlus;

export default [
  // Do not run svgs for now
  // SvgApple,
  // SvgGoogle,
  // SvgFacebook,
  // SvgPlus,
  // SvgSetting,
  // SvgPerson,
  // SvgSearch,
  IC_ADD_W,
  IC_ICON,
  IC_CAMERA,
  IC_CIRCLE_X,
  IC_ERROR,
  IC_PROFILE,
  IC_BACK,
  IC_PIC,
  IC_SEARCH,
  IC_SMILE,
  IC_LOGO_D,
  IC_LOGO_W,
  IC_SETTING_W,
  IC_SEARCH_W,
  IC_PROFILE_W,
];
