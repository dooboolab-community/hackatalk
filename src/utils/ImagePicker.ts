import * as ImagePicker from 'expo-image-picker';
import * as Permissions from 'expo-permissions';

const photoOptions = {
  mediaTypes: ImagePicker.MediaTypeOptions.Images,
  allowsEditing: true,
  exif: true,
};

const requestPermissions = async (
  type: string,
): Promise<Permissions.PermissionStatus> => {
  if (type === 'photo') {
    const { status } = await Permissions.askAsync(Permissions.CAMERA_ROLL);
    return status;
  }
  const { status } = await Permissions.askAsync(Permissions.CAMERA);
  return status;
};

export const launchCameraAsync = async (): Promise<ImagePicker.ImagePickerResult | null> => {
  const permissionStatus = await requestPermissions('camera');
  if (permissionStatus === Permissions.PermissionStatus.GRANTED) {
    const result = await ImagePicker.launchCameraAsync(photoOptions);
    return result;
  }
  return null;
};

export const launchImageLibraryAsync = async (): Promise<ImagePicker.ImagePickerResult | null> => {
  const permissionStatus = await requestPermissions('photo');
  if (permissionStatus === Permissions.PermissionStatus.GRANTED) {
    const result = await ImagePicker.launchImageLibraryAsync(photoOptions);
    return result;
  }
  return null;
};
