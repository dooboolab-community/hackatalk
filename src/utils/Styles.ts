import { Dimensions } from 'react-native';

const { width, height } = Dimensions.get('window');
let calRatio = width <= height ? 16 * (width / height) : 16 * (height / width);

if (width <= height) {
  if (calRatio < 9) {
    calRatio = width / 9;
  } else {
    calRatio = height / 18;
  }
} else {
  if (calRatio < 9) {
    calRatio = height / 9;
  } else {
    calRatio = width / 18;
  }
}

export const screenWidth = width;
export const screenHeight = height;
export const ratio = calRatio / (360 / 9);
