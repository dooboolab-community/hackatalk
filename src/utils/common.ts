export const validateEmail = (email: string): boolean => {
  const re = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return re.test(email);
};

export const validatePassword = (password: string): boolean => {
  const re = /^.*(?=.{6,15})(?=.*[0-9])(?=.*[a-zA-Z])(?=.*[!@#$%^&+=]).*$/;
  return re.test(password);
};
