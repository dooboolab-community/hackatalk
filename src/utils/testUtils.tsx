import React, { ReactElement } from 'react';

import RootProvider from '../providers';
import { ThemeType } from '../types';

export const createTestElement = (
  child: ReactElement,
  themeType?: ThemeType,
): ReactElement => (
  <RootProvider initialThemeType={themeType}>{child}</RootProvider>
);

export const createTestProps = (
  obj: object = {},
  moreScreenProps?: object,
): object | unknown | any => ({
  navigation: {
    navigate: jest.fn(),
    goBack: jest.fn(),
    resetRoot: jest.fn(),
  },
  screenProps: {
    changeThemeType: jest.fn(),
    ...moreScreenProps,
  },
  ...obj,
});
