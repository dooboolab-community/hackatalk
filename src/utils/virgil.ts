import { EThree, ICard, IPrivateKey } from '@virgilsecurity/e3kit-native';

import AsyncStorage from '@react-native-community/async-storage';
import Config from 'react-native-config';

const { ROOT_URL } = Config;

export let eThree: EThree;

/**
 * Virgil Pythia Service has a rate limit of 2 seconds.
 */
const SERVICE_DELAY = 2000;

/**
 * Return a promise that will be resolved after `ms` milliseconds.
 * @param {number} ms - Timeout in milliseconds.
 */
async function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, ms);
  });
}

/**
 * Create a new user in the Virgil Cloud.
 * @param {EThree} e3kit - Instance of EThree class for current user.
 * @param {string} password - User's password.
 */
async function createUser(e3kit: EThree, password: string): Promise<void> {
  try {
    await e3kit.register();
    await e3kit.backupPrivateKey(password);
  } catch (error) {
    if (error.name === ' CloudEntryExistsError') {
      await e3kit.resetPrivateKeyBackup();
      sleep(SERVICE_DELAY);
      await e3kit.backupPrivateKey(password);
    } else {
      throw error;
    }
  }
}

/**
 * Returns true if private key's identifier matches the card's public key identifier.
 * @param {EThree} e3kit - Instance of EThree class for current user.
 * @param {ICard} card - User's Virgil Card.
 */
async function isKeyValid(e3kit: EThree, card: ICard): Promise<boolean> {
  const keyEntry = await e3kit.keyEntryStorage.load(e3kit.identity);

  if (!keyEntry) {
    return false;
  }
  const privateKey: IPrivateKey = e3kit.virgilCrypto.importPrivateKey(keyEntry.value);
  if (privateKey.identifier.byteLength !== card.publicKey.identifier.byteLength) {
    return false;
  }
  for (let i = 0; i < privateKey.identifier.byteLength; i += 1) {
    if (privateKey.identifier[i] !== card.publicKey.identifier[i]) {
      return false;
    }
  }
  return true;
}

/**
 * Rotate user's private key and save it in the Virgil Cloud.
 * @param {EThree} e3kit - Instance of EThree class for current user.
 * @param {string} password - User's password.
 */
async function rotateKey(e3kit: EThree, password: string): Promise<void> {
  await e3kit.cleanup();
  await e3kit.rotatePrivateKey();
  await e3kit.backupPrivateKey(password);
}

export const getTokenFactory = (identity: string): () => Promise<string> => {
  return (): Promise<string> =>
    fetch(`${ROOT_URL}/virgil-jwt/${encodeURIComponent(identity)}`)
      .then((res) => res.text());
};

/**
 * Restore user from the Virgil Cloud.
 * @param {EThree} e3kit - Instance of EThree class for current user.
 * @param {ICard} card - User's Virgil Card.
 * @param {string} password - User's password.
 */
async function restoreUser(e3kit: EThree, card: ICard, password: string): Promise<void> {
  try {
    await e3kit.restorePrivateKey(password);
    const keyValid = await isKeyValid(e3kit, card);
    if (!keyValid) {
      await e3kit.resetPrivateKeyBackup();
      await sleep(SERVICE_DELAY);
      await rotateKey(e3kit, password);
    }
  } catch (error) {
    if (error.name === 'PrivateKeyNoBackupError') {
      await sleep(SERVICE_DELAY);
      await rotateKey(e3kit, password);
    } else {
      throw error;
    }
  }
}

/**
 * Initialize user on current device.
 * @param {EThree} e3kit - Instance of EThree class for current user.
 * @param {string} password - User's password.
 */
export async function initUser(e3kit: EThree, password: string): Promise<void> {
  try {
    await e3kit.cleanup();
    const card = await e3kit.findUsers(e3kit.identity);
    await restoreUser(e3kit, card, password);
  } catch (error) {
    if (error.name === 'UsersNotFoundError') {
      await createUser(e3kit, password);
    } else {
      throw error;
    }
  }
}

export const encryptMessage = async (users: string[]): Promise<string | Buffer> => {
  if (!eThree) {
    throw new Error('eThree is not registered');
  }

  const receiversCard = await eThree.findUsers(users);
  const encryptedMessage = await eThree.encrypt('Hello Receivers!', receiversCard);

  return encryptedMessage;
};

export const decryptMessage = async (users: string[], encryptedMessage: string):
Promise<string | Buffer> => {
  if (!eThree) {
    throw new Error('eThree is not registered');
  }

  const sendersCard = await eThree.findUsers(users);

  const decryptedMessage = await eThree.decrypt(
    encryptedMessage,
    sendersCard[eThree.identity],
  );

  return decryptedMessage;
};

export const initializeEThree = async (userId: string): Promise<void> => {
  const getToken = getTokenFactory(userId);

  try {
    // @ts-ignore ==> https://github.com/VirgilSecurity/virgil-e3kit-js/issues/82
    eThree = await EThree.initialize(getToken, { AsyncStorage });
    const password = await AsyncStorage.getItem('password');
    await initUser(eThree, userId + password);
    AsyncStorage.removeItem('password');
  } catch (err) {
    console.log('Ethree register error', err);
  }
};
