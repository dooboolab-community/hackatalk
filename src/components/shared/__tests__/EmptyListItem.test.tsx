import 'react-native';

import * as React from 'react';

import EmptyListItem from '../EmptyListItem';
import { createTestProps } from '../../../utils/testUtils';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

let props: object;
let component: React.ReactElement;
// let testingLib: RenderResult;

describe('[EmptyListItem] render', () => {
  beforeEach(() => {
    props = createTestProps();
    component = <EmptyListItem {...props} />;
  });

  it('renders without crashing', () => {
    const rendered: renderer.ReactTestRendererJSON | null = renderer
      .create(component)
      .toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  // describe('interactions', () => {
  //   beforeEach(() => {
  //     testingLib = render(component);
  //   });

  //   it('should simulate onPress', () => {
  //     const btn = testingLib.queryByTestId('btn');
  //     act(() => {
  //       fireEvent.press(btn);
  //     });
  //     expect(cnt).toBe(3);
  //   });
  // });
});
