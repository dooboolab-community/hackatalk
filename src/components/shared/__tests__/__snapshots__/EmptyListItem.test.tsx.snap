// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[EmptyListItem] render renders without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "flex": 1,
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "color": "rgb(155,155,155)",
        "fontSize": 14,
      }
    }
  />
</View>
`;
